ip-whitelist/cmd/resource/resource.go

```diff
@@ -0,0 +1,247 @@
+	return strings.Join(encodedValues, "-")
+}
+
+// Read handles the Read event from the Cloudformation service.
+func Read(req handler.Request, prevModel *Model, currentModel *Model) (handler.ProgressEvent, error) {
+	client, err := util.CreateMongoDBClient(*currentModel.ApiKeys.PublicKey.Value(), *currentModel.ApiKeys.PrivateKey.Value())
+	if err != nil {
+		return handler.ProgressEvent{}, err
+	}
+
+	ids := decodeStateID(*currentModel.Id.Value())
+
+	whitelist, err := getProjectIPWhitelist(ids, client)
+	if err != nil {
+		return handler.ProgressEvent{}, err
+	}
+
+	currentModel.Whitelist = flattenWhitelist(whitelist)
+
+	return handler.ProgressEvent{
+		OperationStatus: handler.Success,
+		Message:         "Read Complete",
+		ResourceModel:   currentModel,
+	}, nil
+}
+
+func flattenWhitelist(whitelist []*mongodbatlas.ProjectIPWhitelist) []WhitelistDefinition {
+	var results []WhitelistDefinition
+	for _, wl := range whitelist {
+		r := WhitelistDefinition{
+			IpAddress: encoding.NewString(wl.IPAddress),
+			CidrBlock: encoding.NewString(wl.CIDRBlock),
+			Comment:   encoding.NewString(wl.Comment),
+			ProjectId: encoding.NewString(wl.GroupID),
+		}
+		results = append(results, r)
+	}
+
+	return results
+}
+
+func decodeStateID(stateID string) map[string]string {
+	decode := func(d string) string {
+		decodedString, err := base64.StdEncoding.DecodeString(d)
+		if err != nil {
+			log.Printf("[WARN] error decoding state ID: %s", err)
+		}
+		return string(decodedString)
+	}
+	decodedValues := make(map[string]string)
+	encodedValues := strings.Split(stateID, "-")
+
+	for _, value := range encodedValues {
+		keyValue := strings.Split(value, ":")
+		decodedValues[decode(keyValue[0])] = decode(keyValue[1])
+	}
+	return decodedValues
+}
+
+func getProjectIPWhitelist(ids map[string]string, conn *mongodbatlas.Client) ([]*mongodbatlas.ProjectIPWhitelist, error) {
+	projectID := ids["project_id"]
+	entries := strings.Split(ids["entries"], ",")
+
+	var whitelist []*mongodbatlas.ProjectIPWhitelist
+	for _, entry := range entries {
+		res, _, err := conn.ProjectIPWhitelist.Get(context.Background(), projectID, entry)
+		if err != nil {
+			return nil, fmt.Errorf("error getting project IP whitelist information: %s", err)
+		}
+		whitelist = append(whitelist, res)
+	}
+	return whitelist, nil
+}
+
+// Update handles the Update event from the Cloudformation service.
+func Update(req handler.Request, prevModel *Model, currentModel *Model) (handler.ProgressEvent, error) {
+	client, err := util.CreateMongoDBClient(*currentModel.ApiKeys.PublicKey.Value(), *currentModel.ApiKeys.PrivateKey.Value())
+	if err != nil {
+		return handler.ProgressEvent{}, err
+	}
+
+	ids := decodeStateID(*currentModel.Id.Value())
+
+	whitelist, err := getProjectIPWhitelist(ids, client)
+	if err != nil {
+		return handler.ProgressEvent{}, err
+	}
+
+	whiteListMap(whitelist, func(entry string) {
+		_, err = client.ProjectIPWhitelist.Delete(context.Background(), ids["project_id"], entry)
```

Comment:
Maybe I missed the larger discussion of this that happened in another repo. Why do we opt to manually delete + recreate the whitelist entry, rather than rely on the update behavior of https://docs.atlas.mongodb.com/reference/api/whitelist-update-one/ ?

Reply:
We ran into this first with Terraform.  Basically update depends on the underlying server (Atlas) to figure out what is new and what isn't so any external tool that also needs to track state may not be able to do so within that tools construct.   It's important to ensure we limit the scope of this  - in Terraform we are breaking up the IP whitelist resource to be individual entries and not the list. We have an internal ticket regarding the lack of a true update resource on this too @shum.

Reply:
Understood, thanks for the context!